docs/rfc/rfc-020-block-structure-consolidation.md

```diff
@@ -0,0 +1,363 @@
+and `GasUsed` with the aim of ensuring that execution of all of the transactions
+was performed identically on each node. The data covered by this field _should_
+be also reflected in the `AppHash`. The `AppHash` is provided by the application
+and should be deterministically calculated by each node. This field could
+therefore be removed on the grounds that its data is already reflected elsewhere.
+
+I would advocated for keeping this field. This field provides an additional check
+for determinism across nodes. Logic to update the application hash is more
+complicated for developers to implement because it relies either on building a
+complete view of the state of the application data. The `Results` returned by
+the application contain simple response codes and deterministic data bytes.
+Leaving the field will allow for transaction execution issues that are not
+correctly reflected in the `AppHash` to be more completely diagnosed.
+
+Take the case of mismatch of `LastResultsHash` between two nodes, A and B, where both
+nodes report divergent values. If `A` and `B` both report
+the same `AppHash`, then some non-deterministic behavior occurred that was not
+accurately reflected in the `AppHash`. The issue caused by this
+non-determinism may not show itself for several blocks, but catching the divergent
+state earlier will improve the chances that a chain is able to recover.
+
+#### ValidatorsHash
+
+Both `ValidatorsHash` and `NextValidatorsHash` are included in the block
+header. `Validatorshash` contains the hash of the [public key and voting power][val-hash]
+of each validator in the active set for the current block and `NextValidatorsHash`
+contains the same data but for the next height.
+
+This data is effectively redundant. Having both values present in the block
+structure is helpful for light client verification. The light client is able to
+easily determine if two sequential blocks used the same validator set by querying
+only one header.
+
+`ValidatorsHash` is also important to the light client algorithm for performing block
+validation. The light client uses this field to ensure that the validator set
+it fetched from a full node is correct. It can be sure of the correctness of
+the retrieved structure by hashing it and checking the hash against the `ValidatorsHash`
+of the block it is verifying. Because a validator that the light client trusts 
+signed over the `ValidatorsHash`, it can be certain of the validity of the
+structure. Without this check, phony validator sets could be handed to the light
+client and the code tricked into believing a different validator set was present
+at a height, opening up a major hole in the light client security model.
+
+This creates a recursive problem. To verify the validator set that signed the
+block at height `H`, what information do we need? We could fetch the
+`NextValidatorsHash` from height `H-1`, but how do we verify that that hash is correct?
```

Comment:
In the verify non-adjacent case, we're trying to verify a block of height `H`. To do so, we try to expand the validator set that we trust to include enough of the validators that signed `H`. When we need to perform non-adjacent verification, we fetch a block header along with the commit that supposedly corresponds to that header. Once we verify this header was signed by enough of the validators that we trust, we update the trusted set of validators to include the other signers of that header. 

If we remove the `ValidatorsHash`, it seems that we open ourselves to the following issue:
- We have a known good validator set, `vs1`, and attempt to use it to verify a non-adjacent header.
- We fetch the non-adjacent header along with the untrusted validator set, `vs2`, and the commit, `c`, that justifies this header.
- We check that  `TrustLevel` of `vs1` has signatures in `c` for the header and that `vs2` also has signatures in `c` for the header.
- We now trust `vs2`. However, nothing in `vs1` actually signed off on the veracity of `vs2`. 

A malicious node could compose a `c` and `vs2` that comports with `vs1` so that the client now trusts phony validators in `vs2`. With `ValidatorsHash` in the header, we have reason to trust `vs2`: enough of `vs1` signed off on its hash in the header. Maybe there's a reason this isn't a problem?

Reply:
Do we do this currently? i.e. check hash(`vs2`) = `h2.ValidatorsHash`

Reply:
Yes:
https://github.com/tendermint/tendermint/blob/b0fe38c24543ed61796a865fcdfc34ff57e97ba0/light/verifier.go#L266

Called From:
https://github.com/tendermint/tendermint/blob/b0fe38c24543ed61796a865fcdfc34ff57e97ba0/light/verifier.go#L61

Reply:
Ok yeah that makes sense. Perhaps then we do need to keep `ValidatorsHash` around. Thanks for investigating

Reply:
I'll take a closer look at the light client algorithm definition to be certain and add a description to this doc.

Reply:
I'd be interested in a whiteboard session during the retreat on the subject of how the validator set evolves, and how it's used for (light-client) verification